concurrency/ExpandableListAdapter.java

```diff
@@ -0,0 +1,154 @@
+            highlightedContainer.setContainerHighlighted(false);
+            highlightedContainer = null;
+        }
+        if(containerIndex != UNINITIALIZED) {
+            highlightedContainer = containers.get(containerIndex);
+            highlightedContainer.setContainerHighlighted(true);
+        }
+        if(highlightedContainer != activeContainer && highlightedContainer != null) {
+            highlightActiveMarker(UNINITIALIZED);
+        }
+    }
+
+    public void highlightActiveMarker(int markerIndex) {
+        highlightMarker(activeMarkers, markerIndex);
+    }
+
+    public void highlightMarker(List<ContainerItem> markers, int markerIndex) {
+        if(highlightedMarker != null) {
+            highlightedMarker.setLabelHighlighted(false);
+        }
+        if(markers == null || markerIndex < 0 || markerIndex>=markers.size()) {
+            highlightedMarker = null;
+            return;
+        }
+        ContainerItem marker = markers.get(markerIndex);
+
+        marker.setLabelHighlighted(true);
+        highlightedMarker = marker;
+    }
+
+    public void highlightElement(int containerIndex, int markerIndex) {
+        if(containerIndex == UNINITIALIZED) {
+            highlightMarker(null,markerIndex);
+        } else {
+            highlightMarker((List<ContainerItem>) containers.get(containerIndex).getSubItems(),markerIndex);
+        }
+        highlightContainer(containerIndex);
+        notifyDataSetChanged();
+    }
+
+    public synchronized void clearList() {
+        containers.clear();
+        activeMarkers.clear();
+        notifyDataSetChanged();
+    }
+
+    @Override
+    public synchronized int getCount() {
+        return containers.size() + activeMarkers.size();
+    }
+
+    public synchronized int getCountWithoutPadding() {
+        return containers.size() + activeMarkers.size();
+    }
+
+    @Override
+    public synchronized ExpandableListComponent getItem(int index) {
+        GLogger.d(TAG, "index: %d,  ActiveContainerIndex: %d,  containerSize: %d,  markerSize: %d", index, activeContainerIndex, containers.size(), activeMarkers.size());
+
+        if(activeContainerIndex == UNINITIALIZED || index <= activeContainerIndex) {
+            return containers.get(index);
+        } else if(index > activeContainerIndex + activeMarkers.size()) {
+            return containers.get(index - activeMarkers.size());
+        } else {
+            return activeMarkers.get(index - (activeContainerIndex+1));
+        }
+    }
+
+    @Override
+    public synchronized long getItemId(int index) {
+        return getItem(index).hashCode();
+    }
+
+    public abstract View getView(int i, View view, ViewGroup viewGroup);
+
+    public int getActiveContainerIndex() {
+        return activeContainerIndex;
+    }
+
+    public int containerToListIndex(Integer containerIndex) {
+        if(activeContainerIndex<containerIndex) {
+            return containerIndex+activeMarkers.size();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 This line may not be thread-safe: **146**. At this line the following field is not protected by synchronization: `activeMarkers`. Your code typically protects this field with synchronization on object: `ExpandableListAdapter.this`.

**Fix**  
 Consider protecting the above field with synchronization on object: `ExpandableListAdapter.this`. For example, like at lines: **113**, **117**, **60**, and **107**.

**More info**  
 The following lines are where the above field is already protected by synchronization on the above object: **53**, **54**, **60**, **107**, **113**, **117**, **122**, **126**, **127**, and **129**. This is a call to `size()`. You are already protecting `size()`. You are protecting these operations at lines: **113** and **117**.